File: mobile/ios/TikTokLiveCompanion/BridgeValidator.swift
import Foundation

enum BridgeValidationError: Error, Equatable { case origin, frame, oversized, malformed, unsupported }

struct BridgeValidator {
    static let allowedOrigin = "https://www.tiktok.com"
    static let maximumBytes = 64 * 1024
    static let allowedTypes: Set<String> = [
        "bridge-ready", "inspection", "capability", "chat", "caption", "live-stats", "gift",
        "bridge-error", "command-result", "audio-chunk", "audio-complete"
    ]

    static func decode(data: Data, origin: String, isMainFrame: Bool) throws -> BridgeEnvelope {
        guard origin == allowedOrigin else { throw BridgeValidationError.origin }
        guard isMainFrame else { throw BridgeValidationError.frame }
        guard data.count <= maximumBytes else { throw BridgeValidationError.oversized }
        guard let envelope = try? JSONDecoder().decode(BridgeEnvelope.self, from: data), envelope.version == 1 else { throw BridgeValidationError.malformed }
        guard allowedTypes.contains(envelope.type) else { throw BridgeValidationError.unsupported }
        return envelope
    }

    static func validatedHTTPS(_ value: String?) -> URL? {
        guard let value, let url = URL(string: value), url.scheme == "https" else { return nil }
        return url
    }
}

File: mobile/android/app/src/main/java/app/tiktoklivecompanion/BridgeValidator.kt
package app.tiktoklivecompanion

import android.net.Uri
import org.json.JSONObject

object BridgeValidator {
    const val ALLOWED_ORIGIN = "https://www.tiktok.com"
    const val MAX_BYTES = 64 * 1024
    private val allowedTypes = setOf("bridge-ready", "inspection", "capability", "chat", "caption", "live-stats", "gift", "bridge-error", "command-result", "audio-chunk", "audio-complete", "socket-open", "force-start", "force-return", "player-state", "media-url")

    // Subframes derselben Origin sind erlaubt: TikTok kann den Webcast-WebSocket in einem Same-Origin-Iframe öffnen (0PE-52).
    fun decode(raw: String, origin: String, isMainFrame: Boolean = true): BridgeEnvelope? {
        if (origin != ALLOWED_ORIGIN || raw.toByteArray().size > MAX_BYTES) return null
        return runCatching {
            val json = JSONObject(raw)
            val type = json.getString("type")
            if (json.getInt("version") != 1 || type !in allowedTypes) return null
            val payloadJson = json.optJSONObject("payload") ?: JSONObject()
            val payload = payloadJson.keys().asSequence().associateWith { key -> payloadJson.opt(key).takeUnless { it === JSONObject.NULL } }
            BridgeEnvelope(1, type, json.optString("streamId"), json.getLong("sequence"), json.getString("timestamp"), payload)
        }.getOrNull()
    }

    fun safeHttpsUrl(value: String?): Uri? = value?.let(Uri::parse)?.takeIf { it.scheme == "https" && !it.host.isNullOrBlank() }
}
